vllm/v1/executor/abstract.py

```diff
@@ -0,0 +1,53 @@
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Optional, Set, Tuple
+
+from vllm.config import VllmConfig
+from vllm.v1.outputs import ModelRunnerOutput
+
+
+@dataclass
+class RPCParams:
+    """Arguments for a collective RPC to run on workers"""
+    method: str  # Which Worker method should be executed?
+    output_ranks: Set[int]  # Which output ranks should return to the executor?
+    timeout: Optional[float] = None
+
+
+class Executor(ABC):
+    """Abstract class for executors."""
+
+    @abstractmethod
+    def __init__(self, vllm_config: VllmConfig) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def initialize(self, num_gpu_blocks: int) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def determine_num_available_blocks(self) -> Tuple[int, int]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def execute_model(
+        self,
+        scheduler_output,
+    ) -> ModelRunnerOutput:
+        raise NotImplementedError
+
+    @abstractmethod
+    def profile(self, is_start=True):
+        raise NotImplementedError
+
+    @abstractmethod
+    def shutdown(self):
+        pass
+
+    @abstractmethod
+    def check_health(self) -> None:
+        raise NotImplementedError
+
+    @abstractmethod
+    def collective_rpc(self, rpc: RPCParams, *args, **kwargs) -> [Optional]:
```

Comment:
after a second thought, I think `output_ranks` are not necessary. we should:

```python
def collective_rpc(self, method: str, timeout: Optional[float]=None, *args, **kwargs) -> List[Any]:
    pass
```

`collective_rpc` should check all workers return correctly, collect all the outputs, and 

```python
    def execute_model(
        self,
        scheduler_output,
    ) -> ModelRunnerOutput:
        return self.collective_rpc(method="execute_model", scheduler_output)[-1]
```

some rpc calls can filter out the objects they want, depending on the semantic.

Reply:
What advantage do you see over the current design? If we do this then each worker process will serialize and send the model output regardless of if it's needed. And the current message queue implementation means that the executor will have to deserialize all of them

Reply:
> If we do this then each worker process will serialize and send the model output regardless of if it's needed.

you don't need to send the model output. for non-zero rank workers, even if it has the model output, it should send a `None` object as a placeholder.

in the worker side, it should be:

```python
def execute_model(self, scheduler_output) -> Optional[ModelRunnerOutput]:
    output = self.model_runner(scheduler_output)
    if self.rank != 0:
        return None
    return output
```

The idea is that all workers need to send **something**, to indicate that it executes successfully. otherwise, even if you get the output from rank 0, you don't know if other workers have errors (in which case you should not return the output, but should raise errors).

Reply:
why is this resolved?

Reply:
not sure but definitely should be resolved now